src/Client/ClientBaseHelpers.cpp

```diff
@@ -413,7 +413,42 @@ void highlight(const String & query, std::vector<replxx::Replxx::Color> & colors
 
 String highlighted(const String & query, const Context & context)
 {
-    size_t num_code_points = UTF8::countCodePoints(reinterpret_cast<const UInt8 *>(query.data()), query.size());
+    // Issue: https://github.com/ClickHouse/ClickHouse/issues/83987
+    /// Previously utf-8 code points were calculated in the following way:
+    /// size_t num_code_points = UTF8::countCodePoints(reinterpret_cast<const UInt8 *>(query.data()), query.size());
+    /// But, `UTF8::countCodePoints` and `UTF8::seqLength` seem to handle invalid UTF-8 sequences inconsistently
+    /// (e.g., hex literals like x'A0'), causing count mismatches and crashes.
+    /// For a quick fix, since the highlight function uses `UTF8::seqLength` for iteration, use the same logic for
+    /// counting to ensure consistency when invalid UTF-8 bytes are detected (use UTF8::countCodePoints for all other
+    /// cases to mainly for performance concerns).
+    /// TODO: @bharatnc Fix UTF8::countCodePoints to handle invalid UTF-8 sequences consistently with seqLength so that
+    /// this logic can be removed.
+    bool has_invalid_utf8 = false;
+    for (const char c : query)
+    {
+        /// Standalone UTF-8 continuation bytes (0x80-0xBF, e.g. 0xA0 from hex literals)
+        /// cause countCodePoints/seqLength inconsistency.
+        if (static_cast<unsigned char>(c) >= 0x80 && static_cast<unsigned char>(c) <= 0xBF)
```

Comment:
Not sure that two loops instead of one will be faster, but I'm fine with it anyway